### PR TITLE
NN-3076 Compute running balance (WIP)

### PR DESCRIPTION
**Pull request by JamieDraperUK**
- Compute running balance
- Refactor tests into nested parts

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTransactionHistory.java [hunk 1]
```diff
@@ -78,6 +90,32 @@
     @Default
     private List<OffenderTransactionDetails> relatedTransactionDetails = new ArrayList<>();
 
+    @Builder.Default
+    @Transient
+    private BigDecimal currentBalance = BigDecimal.ZERO;
+
+    @Column(name = "CREATE_DATETIME")
+    @CreatedDate
+    private LocalDateTime createDatetime;
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTransactionHistory.java:99]: The createDatetime property in AuditableEntity was set to protected scope so I couldn't access it via the builder. I tried creating a custom getter and setter that would delegate to super.createDateTime but the Lombok builder still wouldn't pick it up. So in the end I moved over the audit props and exposed the one I need.

Comment (sp-watson) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTransactionHistory.java:99]: Well, I guess you tried it but could we add a method with a different name, e.g:
    public LocalDateTime getCreatedDateTime() {
        return getCreateDatetime();
    }

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java [hunk 2]
```diff
@@ -51,31 +55,71 @@ public OffenderTransactionHistoryService(@Value("${api.currency:GBP}") final Str
...
-            .flatMap(AccountCode::byCodeName)
-            .map(optionalCode -> optionalCode.code)
-            .orElse(null);
-
-        return historyRepository.getTransactionHistory(offenderNo, accountCodeValue, fromDate, toDate, transactionType)
+        return getAllTransactionsWithRunningBalance(offenderNo)
             .stream()
-            .sorted(SORT_BY_MOST_RECENT_DATE_THEN_BY_LATEST_SEQ)
+            .sorted(SORT_BY_RECENT_DATE)
+            .filter(byDateRange(fromDate, toDate))
```
[the remark this answers is not in the document]
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java:60]: It certainly wouldn't hurt.

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTransactionHistory.java (hunk 1 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTransactionHistory.java:99]: I did try that but the new getter wouldn't seem to conform to the Lombok builder interface.  Maybe I messed it up.

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 3]
```diff
@@ -41,301 +40,398 @@ public void setUp() {
...
     }
 
-    @Test
-    public void testGetTransactionHistory_CallsRepositoryWithCorrectParameters() {
+    @Nested
+    public class CorrectParameters {
+        @Test
+        public void testGetTransactionHistoryThrowException_offenderIdIsNull() {
+
+            Throwable exception = assertThrows(NullPointerException.class, () -> {
```
Comment (mikehalmamoj) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:48]: AssertJ's `assertThatThrownBy` is worth considering in future - has some of the usual AssertJ goodness.

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java [hunk 4]
```diff
@@ -41,301 +40,398 @@ public void setUp() {
...
+            assertThat(histories).isNotNull();
+            assertThat(histories.size()).isEqualTo(9);
 
-        verify(repository, times(1)).getTransactionHistory(OFFENDER_NO, "SPND", fromDate, null, transactionType);
 
-        assertThat(histories).isNotNull();
-        assertThat(histories.size()).isEqualTo(0);
-    }
+            // Date: now, Seq: 3, 2, 1
+            assertThat(histories.get(0).getTransactionId()).isEqualTo(1L);
```
Comment (mikehalmamoj) [inline src/test/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryServiceTest.java:205]: Maybe some `extracting` and `containsExactly` might make this a little less verbose?

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**